### PR TITLE
Fix sign issues

### PR DESCRIPTION
**Pull request by ichetty**
C_Sign is called twice, first with null buffer to get the buffer size. This changes will handle the first case and return the required buffer size.

### CODE REVIEW

**Comment by williamcroberts** [comment]
You're patches should never introduce whitespace errors.
```
.git/rebase-apply/patch:71: trailing whitespace.
            if not exists:   
.git/rebase-apply/patch:78: trailing whitespace.
          
warning: 2 lines add whitespace errors.
```

File: tools/tpm2_ptool.py [hunk 1]
```diff
@@ -891,7 +941,7 @@ def do_token_init(db, path, args):
 
             if args['wrap'] != 'auto':
                 if args['wrap'] == 'software' and sym_support:
-                    print("Warning: confifuring software wrapping key when TPM has support.\n"
+                    print("Warning: configuring software wrapping key when TPM has support.\n"
```
Comment (williamcroberts) [inline tools/tpm2_ptool.py]: You could split this out to a separate patch. `git add -p` is your friend.

File: tools/tpm2_ptool.py [hunk 2]
```diff
@@ -949,7 +999,7 @@ def do_token_init(db, path, args):
                 encsobjauth = AESCipher(wrappingobjauth['rhash']).encrypt(sobjauth)
 
             objattrs="restricted|decrypt|fixedtpm|fixedparent|sensitivedataorigin|userwithauth"
-            sobjpriv, sobjpub, sobjpubdata = tpm2.create(pobject['handle'], pobjauthhash, sobjauth, objattrs=objattrs, alg='aes{}'.format(sym_size))
+            sobjpriv, sobjpub, sobjpubdata = tpm2.create(pobject['handle'], pobjauthhash, sobjauth, objattrs=objattrs, alg='rsa2048')
```
Comment (williamcroberts) [inline tools/tpm2_ptool.py]: Isn't rsa2048 the default?

File: tools/tpm2_ptool.py [hunk 3]
```diff
@@ -1233,6 +1283,206 @@ def __call__(self, args):
 
                 print('Added key as label: "{keylabel}"'.format(keylabel=keylabel))
 
+@commandlet("importrsa")
+class ImportRsaCommand(Command):
```
Comment (williamcroberts) [inline tools/tpm2_ptool.py]: just a commandlet called *import* or *importkey* with an option for specifying algorithm. Just like *addkey* command. I wonder if it would also be advantageous to have a *---import* option to the addkey command.... probably not.

File: tools/tpm2_ptool.py [hunk 4]
```diff
@@ -1233,6 +1283,206 @@ def __call__(self, args):
...
+            required=True)
+        group_parser.add_argument(
+            '--label',
+            help='The tokens label to add a key too.\n',
+            required=True)
+        group_parser.add_argument(
+            '--key-label',
+            help='The label of the key imported.\n')
+        group_parser.add_argument(
+            '--algorithm',
```
Comment (williamcroberts) [inline tools/tpm2_ptool.py]: see you have it already

File: tools/tpm2_ptool.py [hunk 5]
```diff
@@ -1233,6 +1283,206 @@ def __call__(self, args):
...
+        pinopts.add_argument(
+            '--sopin',
+            help='The Administrator pin,'),
+        pinopts.add_argument(
+            '--userpin',
+            help='The User pin.\n'),
+
+    def __call__(self, args):
+
+
```
Comment (williamcroberts) [inline tools/tpm2_ptool.py]: ETOOMANY newlines

File: tools/tpm2_ptool.py [hunk 6]
```diff
@@ -1233,6 +1283,206 @@ def __call__(self, args):
...
+                if is_so:
+                    pin = args['sopin']
+                    pinpobjauthkeysalt = token['sopobjauthkeysalt']
+                    pinpobjauthkeyiters = token['sopobjauthkeyiters']
+                    pinpobjauth = token['sopobjauth']
+                else:
+                    pin = args['userpin']
+                    pinpobjauthkeysalt = token['userpobjauthkeysalt']
+                    pinpobjauthkeyiters = token['userpobjauthkeyiters']
+                    pinpobjauth = token['userpobjauth']
```
Comment (williamcroberts) [inline tools/tpm2_ptool.py]: This duplicates addkey command a lot. Refactor this so common blocks are moved and shared between the two. Perhaps use class inheritance and provide just the parts that differ....

I think you would be able to do a super class called generickey or something like that, and then have each addkey and importkey commandlets inherit that generickey class. The call could call the super class and then add the missing functionality via some interface.

They both do the same thing except one does inport and the other calls create. See how I did *AddEmptyTokenCommand* command.

File: test/integration/scripts/create_pkcs_store.sh [hunk 7]
```diff
@@ -69,8 +69,8 @@ tpm2_ptool.py addtoken --pid=2 --pobj-pin=anotherpobjpin --sopin=anothersopin --
 # add 2 aes and 2 rsa keys under tokens 1 and 2
 for t in "label" "wrap-sw"; do
 	echo "Adding 2 AES 256 keys under token \"$t\""
-	tpm2_ptool.py addkey --algorithm=aes256 --label="$t" --userpin=myuserpin --path=$TPM2_PKCS11_STORE
-	tpm2_ptool.py addkey --algorithm=aes256 --label="$t" --key-label=mykeylabel --userpin=myuserpin --path=$TPM2_PKCS11_STORE
+	tpm2_ptool.py addkey --algorithm=aes128 --label="$t" --userpin=myuserpin --path=$TPM2_PKCS11_STORE
+	tpm2_ptool.py addkey --algorithm=aes128 --label="$t" --key-label=mykeylabel --userpin=myuserpin --path=$TPM2_PKCS11_STORE
```
[the remark this answers is not in the document]
Comment (ichetty) [reply test/integration/scripts/create_pkcs_store.sh]: yes, but the discrete tpm does not support. to have this script work on both simulator and tpm I changed it to aes128. We could add a script to check that's the max supported just like its done in the tool.

File: src/lib/sign.c [hunk 8]
```diff
@@ -278,7 +287,10 @@ CK_RV sign_final (CK_SESSION_HANDLE session, unsigned char *signature, unsigned
 
 CK_RV sign(CK_SESSION_HANDLE session, unsigned char *data, unsigned long data_len, unsigned char *signature, unsigned long *signature_len) {
 
-    CK_RV rv = sign_update(session, data, data_len);
+	CK_RV rv = CKR_OK;
```
[the remark this answers is not in the document]
Comment (ichetty) [reply src/lib/sign.c]: Got it!

File: test/integration/scripts/create_pkcs_store.sh [hunk 9]
```diff
@@ -56,7 +56,7 @@ set -e
 tpm2_ptool.py init --pobj-pin=mypobjpin --path=$TPM2_PKCS11_STORE
 
 # Test the existing primary object init functionality
-tpm2_createprimary -p foopass -o primary.ctx
+tpm2_createprimary -p foopass -o primary.ctx -g sha256 -G rsa
```
[the remark this answers is not in the document]
Comment (ichetty) [reply test/integration/scripts/create_pkcs_store.sh]: Could be defaults I'm not sure. without this change, tpm2_createprimary failed on the discrete tpm so I had to add this here

File: test/integration/scripts/create_pkcs_store.sh [hunk 10]
```diff
@@ -83,6 +83,14 @@ done;
 # add 1 aes key under label "import-primary"
 tpm2_ptool.py addkey --algorithm=aes128 --label="import-primary" --userpin=anotheruserpin --path=$TPM2_PKCS11_STORE
 
+# import 1 rsa2048 key under label "import-primary"
+echo "importing rsa2048 key under token 'import-primary'"
+openssl genrsa -out private.pem 2048
+tpm2_ptool.py importrsa --privkey='private.pem' --algorithm=rsa --key-label="imported_key" --label="import-primary" --userpin=anotheruserpin --path=$TPM2_PKCS11_STORE
+rm -f private.pem
+
+rm -f primary.ctx
```
[the remark this answers is not in the document]
Comment (ichetty) [reply test/integration/scripts/create_pkcs_store.sh]: I'll add the trap. I'm deleting two different files. I saw that the primary.ctx that was created above was not deleted.

File: tools/tpm2_ptool.py (hunk 2 above)
Comment (ichetty) [reply tools/tpm2_ptool.py]: I'm not sure, will have to check the code.

File: tools/tpm2_ptool.py (hunk 3 above)
Comment (ichetty) [reply tools/tpm2_ptool.py]: I'll rename it to import. I added rsa because that's only algorithm import currently supports.

File: tools/tpm2_ptool.py (hunk 6 above)
Comment (ichetty) [reply tools/tpm2_ptool.py]: Definitely, just wanted to check if this worked given the urgency.

**Comment by williamcroberts** [comment]
> -----Original Message-----
> From: Jay Chetty [mailto:notifications@github.com]
> Sent: Friday, October 26, 2018 4:14 PM
> To: tpm2-software/tpm2-pkcs11 <tpm2-pkcs11@noreply.github.com>
> Cc: Roberts, William C <william.c.roberts@intel.com>; Comment
> <comment@noreply.github.com>
> Subject: Re: [tpm2-software/tpm2-pkcs11] Fix sign issues (#61)
> 
> @ichetty commented on this pull request.
> 
> ________________________________
> 
> In test/integration/scripts/create_pkcs_store.sh <https://github.com/tpm2-
> software/tpm2-pkcs11/pull/61#discussion_r228687408> :
> 
> > @@ -69,8 +69,8 @@ tpm2_ptool.py addtoken --pid=2
> > --pobj-pin=anotherpobjpin --sopin=anothersopin --
>  # add 2 aes and 2 rsa keys under tokens 1 and 2  for t in "label" "wrap-sw"; do
>  	echo "Adding 2 AES 256 keys under token \"$t\""
> -	tpm2_ptool.py addkey --algorithm=aes256 --label="$t" --
> userpin=myuserpin --path=$TPM2_PKCS11_STORE
> -	tpm2_ptool.py addkey --algorithm=aes256 --label="$t" --key-
> label=mykeylabel --userpin=myuserpin --path=$TPM2_PKCS11_STORE
> +	tpm2_ptool.py addkey --algorithm=aes128 --label="$t" --
> userpin=myuserpin --path=$TPM2_PKCS11_STORE
> +	tpm2_ptool.py addkey --algorithm=aes128 --label="$t"
> +--key-label=mykeylabel --userpin=myuserpin --path=$TPM2_PKCS11_STORE
> 
> yes, but the discrete tpm does not support. to have this script work on both
> simulator and tpm I changed it to aes128. We could add a script to check that's the
> max supported just like its done in the tool.

Yeah but there is NO testing against discrete TPMS in the CI. So this reduces test
coverage.

File: test/integration/scripts/create_pkcs_store.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (ichetty) [reply test/integration/scripts/create_pkcs_store.sh]: Removed this change in the new PR.

**Comment by ichetty** [comment]
Closing this one as I submitted another new PR
